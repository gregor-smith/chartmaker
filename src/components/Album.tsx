import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
