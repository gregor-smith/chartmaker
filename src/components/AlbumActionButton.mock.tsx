const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
