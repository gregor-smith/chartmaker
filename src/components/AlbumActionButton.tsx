import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
