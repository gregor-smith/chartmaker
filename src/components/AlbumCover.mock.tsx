import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
