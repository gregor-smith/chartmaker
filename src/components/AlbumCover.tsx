import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
