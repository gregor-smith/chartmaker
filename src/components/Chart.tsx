import type { FC, Ref } from 'react'
import { css } from 'emotion'

import { CONTAINER_PADDING_SIZE, BORDER } from '@/style'


const outContainerStyle = css({
    display: 'flex',
    flexDirection: 'column',
    padding: CONTAINER_PADDING_SIZE,
    border: BORDER
})


const innerContainerStyle = css({
    display: 'flex',
    alignItems: 'flex-start'
})


export type ChartProps = {
    name: string
    innerRef: Ref<HTMLElement>
}


export const Chart: FC<ChartProps> = ({ name, innerRef, children }) =>
    <main ref={innerRef} className={outContainerStyle}>
        <h1>{name}</h1>
        <div className={innerContainerStyle}>
            {children}
        </div>
    </main>
