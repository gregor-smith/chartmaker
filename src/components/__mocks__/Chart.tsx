import type { ChartProps } from '@/components/Chart'


export function Chart<TNamedAlbum, TPlaceholderAlbum>({
    albums,
    name,
    rowsX,
    rowsY,
    shape,
    albumRowComponent: AlbumRowComponent,
    titleGroupComponent: TitleGroupComponent
}: ChartProps<TNamedAlbum, TPlaceholderAlbum>) {
    const albumsJSON = JSON.stringify(albums, undefined, 2)
    const shapeJSON = JSON.stringify(shape)
    return (
        <div className='mock-chart'>
            {`Albums: ${albumsJSON}`}
            {`Name: ${name}`}
            {`Rows X: ${rowsX}`}
            {`Rows Y: ${rowsY}`}
            {`Shape: ${shapeJSON}`}
            <AlbumRowComponent albums={[]} size='Mock chart size'/>
            <TitleGroupComponent group={[]}/>
        </div>
    )
}
