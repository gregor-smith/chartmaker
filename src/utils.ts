import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
