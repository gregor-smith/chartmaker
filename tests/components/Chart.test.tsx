import { createRef } from 'react'
import { render } from 'react-dom'

import { Chart } from '@/components/Chart'

import { RenderContainer } from '../utils'


const container = new RenderContainer()



test('renders header and children, passes ref', () => {
    const ref = createRef<HTMLElement>()

    render(
        <Chart innerRef={ref} name='Test chart'>
            Test children
        </Chart>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
    expect(ref.current).toBeInstanceOf(Element)
})


// test.each([ 40, 42, 100 ] as const)('renders with top shape', size => {
//     render(
//         <Chart albums={albums}
//             name={`Test top ${size} chart`}
//             shape={{ tag: 'Top', size }}
//             rowsX={10}
//             rowsY={10}
//             innerRef={ignore}
//             dispatch={ignore}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     expect(container.element).toMatchSnapshot()
// })


// test('renders with collage shape', () => {
//     render(
//         <Chart albums={albums}
//             name='Test 5x7 collage chart'
//             shape={{ tag: 'Collage' }}
//             rowsX={5}
//             rowsY={7}
//             innerRef={ignore}
//             dispatch={ignore}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     expect(container.element).toMatchSnapshot()
// })


// test('excludes placeholder albums from titles', () => {
//     const albumsWithPlaceholders: Album[] = albums.map((album, index) =>
//         index % 2 == 0
//             ? album
//             : album.id
//     )
//     render(
//         <Chart albums={albumsWithPlaceholders}
//             name='Test chart with placeholders'
//             shape={{ tag: 'Collage' }}
//             rowsX={5}
//             rowsY={5}
//             innerRef={ignore}
//             dispatch={ignore}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     expect(container.element).toMatchSnapshot()
// })


// test('passes ref', () => {
//     const ref = createRef<HTMLElement>()

//     render(
//         <Chart albums={albums}
//             name='Test 5x7 collage chart'
//             shape={{ tag: 'Collage' }}
//             rowsX={5}
//             rowsY={7}
//             innerRef={ref}
//             dispatch={ignore}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     expect(ref.current).toBeInstanceOf(Element)
// })


// test('dispatches unhighlight action on mouse leaving albums container', () => {
//     const mock = jest.fn<[], [ Action ]>()

//     render(
//         <Chart albums={albums}
//             name='Test 5x7 collage chart'
//             shape={{ tag: 'Collage' }}
//             rowsX={5}
//             rowsY={7}
//             innerRef={ignore}
//             dispatch={mock}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     act(() => {
//         debugger
//         fireEvent('mouseLeave', container.element?.querySelector(`#${rowsID}`))
//     })

//     expect(mock).toHaveBeenCalledTimes(1)
//     expect(mock).toHaveBeenCalledWith<[ Action ]>({
//         tag: 'UnhighlightAlbum'
//     })
// })


// test('dispatches unhighlight action on mouse leaving titles container', () => {
//     const mock = jest.fn<[], [ Action ]>()

//     render(
//         <Chart albums={albums}
//             name='Test 5x7 collage chart'
//             shape={{ tag: 'Collage' }}
//             rowsX={5}
//             rowsY={7}
//             innerRef={ignore}
//             dispatch={mock}
//             isNamedAlbum={isNamedAlbum}
//             albumRowComponent={TestAlbumRow}
//             titleGroupComponent={TestTitleGroup}/>,
//         container.element
//     )

//     act(() => fireEvent('mouseLeave', container.element?.querySelector(`#${titlesID}`)))

//     expect(mock).toHaveBeenCalledTimes(1)
//     expect(mock).toHaveBeenCalledWith<[ Action ]>({
//         tag: 'UnhighlightAlbum'
//     })
// })
