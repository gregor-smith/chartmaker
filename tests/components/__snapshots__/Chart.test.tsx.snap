// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders header and children, passes ref 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 1rem;
  border: 1px solid white;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

<div
  id="test-render-container"
>
  <main
    class="emotion-1"
  >
    <h1>
      Test chart
    </h1>
    <div
      class="emotion-0"
    >
      Test children
    </div>
  </main>
</div>
`;
